packages/foundry/contracts/hooks/rebalancer/Rebalancer.sol

```diff
@@ -0,0 +1,633 @@
+    address private weightedPoolFactory;
+
+    ///@dev address of the oracle
+    address private oracle;
+    /**
+     * Emmited when the hook is registered successfully for the given pool
+     * @param hooksContract The address of this contract
+     * @param pool  The address of the pool that the hook is used for
+     */
+    event RebalancerHookRegistered(address indexed hooksContract, address indexed pool);
+
+    /**
+     * emitted whenever liquidity is added to the pool
+     * @param pool address of the pool
+     * @param lp the address of liquidity provider
+     */
+    event LiquidtyProviderAdded(address indexed pool, address indexed lp);
+
+    /**
+     * Emmited when the rebalance is triggered for the pool
+     * @param pool The address of the pool
+     */
+    event RebalanceStarted(address indexed pool);
+
+    /**
+     * @notice The hook only supports weighted pools
+     * @dev The math is written for weighted pools only though it can be extended to other pools
+     * but the math will be simpler for CPP or CSP.
+     */
+    error OnlyWeightedPoolsAllowed();
+
+    /**
+     * @notice only pool creator can change rebalancing metrics for pools
+     */
+    error OnlyCreatorCanChangeRebalanceData(address pool);
+
+    /**
+     * @notice The rebalance data is not set for the pool which triggers the rebalancing
+     */
+    error RebalanceDataNotSet(address pool);
+
+    /**
+     * @notice The error given anywhere we encounter dissimilar array lengths
+     * @dev The param indicates various places the error has accoured
+     * example: "PW_N_PT" is Pool weights do not match pool tokens
+     */
+    error InvalidArrayLenghts(string message);
+
+    /**
+     * @notice Hooks functions called from an external router.
+     * @dev This contract inherits both `MinimalRouter` and `BaseHooks`, and functions as is its own router.
+     * @param router The address of the Router
+     */
+    error CannotUseExternalRouter(address router);
+
+    constructor(
+        IVault vault,
+        IPermit2 permi2,
+        IWETH weth,
+        address _weightedPoolFactory,
+        address _oracle
+    ) MinimalRouterWithSwap(vault, weth, permi2) Ownable(msg.sender) {
+        weightedPoolFactory = _weightedPoolFactory;
+        oracle = _oracle;
+    }
+
+    modifier onlySelfRouter(address router) {
+        _ensureSelfRouter(router);
+        _;
+    }
+    ///@dev prevent race conditions when rebalancing
+    bool private isRebalancing;
+
+    modifier nonReentrantRebalance() {
+        require(!isRebalancing, "Rebalance in progress");
+        isRebalancing = true;
+        _;
+        isRebalancing = false;
+    }
+
+    /***************************************************************************
+                                  Router Functions
+    ***************************************************************************/
+    function addLiquidityProportional(
```

Comment:
So if I understand this, you're making people add/remove/swap through this hook (it reverts if you use a standard router). When they add, the tokens go to the Vault, but the hook gets the BPT. When they remove, the hook burns the BPT and gives them the tokens from the Vault. It's tracking the LPs and their token balances inside the hook (whereas the Vault just has the total).

When they swap, it does the normal operation... then potentially redistributes tokens afterward by pulling tokens from LPs and sending back adjusted amounts (fuzzy on how this settles), which seems like it could be done more simply as a "counter-swap".

Reply:
Yeah so I guess the BPT tracking part got a little mixed up it was not my intention to take bpt but to only track the bpt amount to get the active LPs for the pool as I couldn't find the API to do so!

So the problem I tried to solve could be understand as follows. Let's say we have stETH/eth pool:
So as we know when the ethereum network distributes staking rewards the stETH token rebases i.e each token is worth a little more than ETH

So let me take you through an example
The Initial State for the pool might look like:
- 1 stETH = 1 ETH
- LP Pool has 100 ETH and 100 stETH
- Pool value = 200 ETH equivalent
- Expected staking reward: 0.5%
- Post-rebase: 1 stETH will = 1.005 ETH

Now a rebase happens daily so what an arbitrageur can do pre rebase is 
1. Borrow 100 ETH
2. Sell ETH to pool for stETH at 1:1 ratio
and after rebase as the stETH value increases
- stETH in pool rebases to 1.005 ETH each
- Arbitrageur buys back 100. 5 ETH using 100 stETH
- Returns borrowed 100 ETH
- Profit: 0.5 ETH

Therefore the vaule - 0.5 which was the LP's staking reward was lost to arbitrageur.

This would be the case for every dividend/yeild based RWAs pool

So what my plan is to have an oracle tell me that an event like rebase is coming for my pool which is given by the predicted price and when this price is significant my pool rebalncing the liquidity pool so in this case it would take out 0.5 stETH out and protect the LPs from losing out ont the profit

Reply:
And the principal for rebalancing the pool is actually very simple just keep the value same in the pool. I worked out the math and in terms of token changes it would look something like this 
![image](https://github.com/user-attachments/assets/99e4f453-aa9f-4c8c-a677-931755addd76)
Where Bt' is our new balance and rt is the ratio of new price and last price and K are the number of tokens

Reply:
For the counter swap argument I think it can be done for a two token pool. But for multiple token I would have to implement some sort of algorithm that would give to swap which token for which as I can only swap two tokens at  a time (I think). 
So yeah counter swap is better but could get complex with increased number of tokens!